src/components/Swap/Summary/index.tsx

```diff
@@ -27,59 +30,176 @@ const PriceImpactText = styled.span`
+        text={t`Approve token for trading`}
+        content={t`Gives you the ability to trade this token on the Uniswap protocol. For security, this will expire in 30 days.`}
+      />
+    ) : (
+      <PermitTooltipText
+        text={t`Approve permit`}
+        content={t`Permit2 allows safe sharing and management of token approvals across different smart contracts.`}
+      />
+    ),
+    icon: Spinner,
+    onClick: cancel,
+    children: <Trans>Cancel</Trans>,
+  }
+}
+
+enum ReviewState {
+  REVIEWING,
+  ALLOWANCE_PENDING,
+  ALLOWANCE_FAILED,
+  TRADE_CHANGED,
+  PENDING_SWAP,
+}
+
 function ConfirmButton({
   trade,
   onConfirm,
   onAcknowledgeNewTrade,
+  allowance,
 }: {
   trade: InterfaceTrade
   onConfirm: () => Promise<void>
   onAcknowledgeNewTrade: () => void
+  allowance: Allowance
 }) {
+  const [reviewState, setReviewState] = useState(ReviewState.REVIEWING)
+
   const { onSwapPriceUpdateAck, onSubmitSwapClick } = useAtomValue(swapEventHandlersAtom)
   const [ackTrade, setAckTrade] = useState(trade)
   const doesTradeDiffer = useMemo(
     () => Boolean(trade && ackTrade && tradeMeaningfullyDiffers(trade, ackTrade)),
     [ackTrade, trade]
   )
+  useEffect(() => {
+    if (doesTradeDiffer && reviewState === ReviewState.REVIEWING) {
+      setReviewState(ReviewState.TRADE_CHANGED)
+    } else if (!doesTradeDiffer && reviewState === ReviewState.TRADE_CHANGED) {
+      setReviewState(ReviewState.REVIEWING)
+    }
+  }, [reviewState, doesTradeDiffer])
+
+  const isApproved = useMemo(
+    () => (allowance.state === AllowanceState.REQUIRED ? allowance.isApproved : true),
+    [allowance]
+  )
 
-  const [isPending, setIsPending] = useState(false)
-  const onClick = useCallback(async () => {
-    setIsPending(true)
+  const triggerSwap = useCallback(async () => {
+    setReviewState(ReviewState.PENDING_SWAP)
     onSubmitSwapClick?.(trade)
     await onConfirm()
-    setIsPending(false)
+    setReviewState(ReviewState.REVIEWING)
   }, [onConfirm, onSubmitSwapClick, trade])
 
-  const action = useMemo((): Action | undefined => {
-    if (isPending) {
-      return {
-        message: <Trans>Confirm in your wallet</Trans>,
-        icon: Spinner,
-        onClick: () => {
-          setIsPending(false)
-        },
-      }
-    } else if (doesTradeDiffer) {
-      return {
-        color: 'accent',
-        message: <Trans>Price updated</Trans>,
-        icon: AlertTriangle,
-        onClick: () => {
-          onSwapPriceUpdateAck?.(ackTrade, trade)
-          setAckTrade(trade)
-          // Prompts parent to show speedbump if new trade has high impact
-          onAcknowledgeNewTrade()
-        },
-        children: <Trans>Accept</Trans>,
+  const prevAllowanceRef = useRef(allowance)
+  // Ensures swap isn't submitted until allowance has been properly updated post-permit2 flow
+  useEffect(() => {
+    if (prevAllowanceRef.current.state === AllowanceState.REQUIRED && allowance.state === AllowanceState.ALLOWED) {
```

Comment:
hmm this `useEffect` took me a bit to understand - would it be more accurate to have comment update describe this as "automatically trigger signing swap tx if allowance requirements are met"